Simple theme svg2png

dalf opened the pull request:
## What does this PR do?

In the master branch, convert (imagemagick) creates `searx/static/themes/simple/img/favicon.png` from `src/brand/searxng-wordmark.svg`. Some versions of convert have issues to render the .svg file.

This PR replaces the usage of convert by [sharp](https://github.com/lovell/sharp) (a standalone lib, it requires C++ to compile the library).

A custom grunt task is created because [grunt-task](https://www.npmjs.com/package/grunt-sharp) can't be installed (most probably because it hasn't been updated since a long time).

The "[build] /static" has been made using this change (the png file is smaller, it can be even smaller using [optipng](https://www.npmjs.com/package/optipng))

## Why is this change important?

Reliable build steps

## How to test this PR locally?

* `make themes.simple` or `npm run build`
* check `searx/static/themes/simple/img/favicon.png` is not modified (so `make static.build.commit` won't include the favicon at the build).
* check the favicon is rendered as expected.

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

<!--
Closes #234
-->

dalf committed:
[mod] simple theme: use sharp instead of convert to create .png from .svg

define a custom grunt task, since grunt-sharp is too old (it can't be installed).
in gruntfile.js, the image tasks are moved at the end the build chain.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -13,7 +13,7 @@ module.exports = function(grunt) {
       scripts: {
         files: ['gruntfile.js', 'src/**'],
-        tasks: ['eslint', 'copy', 'concat', 'svg2jinja', 'uglify', 'image', 'less:development', 'less:production']
+        tasks: ['eslint', 'copy', 'concat', 'uglify', 'less:development', 'less:production', 'image', 'svg2png', 'svg2jinja']
```
nit: Reordering the items in the tasks list makes it hard to see what has been changed.  Are there any special reason to reorder the tasks?

dalf replied:
Sorry, the purpose is to put all image tasks at the end, where it is easy to list which image is processed.
I should have done that in another PR.

return42 replied:
> Sorry, the purpose is to put all image tasks at the end, where it is easy to list which image is processed. I should have done that in another PR.

No need for a new PR, I can split the one commit in two commits, if you want.

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -240,10 +269,11 @@ module.exports = function(grunt) {
     'stylelint',
     'copy',
     'concat',
-    'svg2jinja',
     'uglify',
-    'image',
     'less:development',
-    'less:production'
+    'less:production',
+    'image',
+    'svg2png',
+    'svg2jinja',
```
see my comment above ..

> Reordering the items in the task list makes it hard to see what has been changed. Are there any special reason to reorder the tasks?

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -100,16 +100,6 @@ module.exports = function(grunt) {
         }
       }
     },
-    image: {
-      svg4web: {
-        options: {
-          svgo: ['--config', 'svg4web.svgo.js']
-        },
-        files: {
-          '<%= _templates %>/__common__/searxng-wordmark.min.svg': '<%= _brand %>/searxng-wordmark.svg'
-        }
-      }
-    },
```
see my comment above ..

> Reordering the items in the task list makes it hard to see what has been changed. Are there any special reason to reorder the tasks?

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -222,6 +228,29 @@ module.exports = function(grunt) {
     grunt.log.ok(this.data.dest + " created");
   });
 
+  grunt.registerMultiTask('svg2png', 'Convert SVG to PNG', function() {
```
Wow, very cool hadn't thought it is such easy :+1:

mrpaulblack reviewed (approved):
LGTM

I tested with `make node.env` and `make themes.simple` and generated new favicon properly and droped it with `make static.build.restore`; also the image is rendered properly in my browser

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -222,6 +228,29 @@ module.exports = function(grunt) {
+  grunt.registerMultiTask('svg2png', 'Convert SVG to PNG', function() {
+    const sharp = require('sharp');
+    for(const pngFileName in this.data.files) {
+      const svgFileName = this.data.files[pngFileName];
```
I don't know what is common, but intuitive I would suggest to change "key" & "value" of the dictionary:  

    files[<src svg>]: "<dest png>" 
 
or even better a list of destinations:

    files[<src svg>]: ["<dest-1 png>", "<dest-2 png> ...] 

EDIT: done in 171f492c

dalf replied:
For these packages picked randomly:
* https://www.npmjs.com/package/grunt-image
* https://www.npmjs.com/package/grunt-svgmin
* https://www.npmjs.com/package/grunt-plantuml

Each time the pattern is:
```
{ "dest": "src" }
```

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -222,6 +228,29 @@ module.exports = function(grunt) {
+            if (err !== null) {
+              grunt.log.error(err);
+            }
+          });
```
the `-resize 64x64` is missed here. (fixed in 171f492c)

return42 reviewed (changes requested):
> it requires C++ to compile the library

Not on a _common_ build host, [Ready-compiled sharp and libvips binaries are provided for use on the most common platforms ... An ~7MB tarball containing libvips and its most commonly used dependencies is downloaded via HTTPS and stored within node_modules/sharp/vendor during npm install.](https://sharp.pixelplumbing.com/install#prebuilt-binaries)

> The "[build] /static" has been made using this change (the png file is smaller, it can be even smaller using optipng)

I missed the resize to 64x64 .. 5096 Byte --> 2747 Byte

I added commit 171f492c on top of your branch ...

    [mod] improve grunt task: svg2png / add resize options

.. it is my first Node.js implementation, please take a thorough look at it.

And I had some nits about reordering the tasks / see below.

dalf commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -137,6 +127,30 @@ module.exports = function(grunt) {
+        files: [
+          {
+            src: '<%= _brand %>/searxng-wordmark.svg',
+            dest: [ 'img/favicon.png' ]
```
The pattern
```
        files: {
          'img/favicon.png': '<%= _brand %>/searxng-wordmark.svg'
        }
```
make it easy to read that `'<%= _brand %>/searxng-wordmark.svg'` is used in this task and the one just above.

return42 committed:
[mod] gruntfile: do some polish (no functional change)

- remove unused 'path' constant
- fix indentation
- Put all image tasks at the end, where it is easy to list which image is processed.

Suggested -by: @dalf https://github.com/searxng/searxng/pull/541#discussion_r756247157
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 reviewed (approved):
I separated ..

64228b69 [mod] gruntfile: do some polish (no functional change) 

from my side this PR is ready to merge / @dalf if don't see any issue more, please merge.

return42 committed:
[build] /static

return42 reviewed (approved):
FYI: squashed the two commits into 59f4c792b and rebased on current master.
